gnomad/utils/constraint.py

```diff
@@ -360,3 +362,76 @@ def collapse_strand(
         if isinstance(t, hl.Table)
         else t.annotate_rows(**collapse_expr)
     )
+
+
+def annotate_constraint_groupings(
```

Comment:
I'm wondering if this should be split into two functions.
One that is just the explode, so sort of like this:
```
def explode_by_vep_annotation(
    t: Union[hl.Table, hl.MatrixTable],
    vep_annotation: str,
    vep_root: str = "vep",
) -> Tuple[Union[hl.Table, hl.MatrixTable], Tuple[str]]:
    # Annotate 'worst_csq_by_gene' to t if t is used to build the "worst_csq" model.
    if vep_annotation == "worst_csq_by_gene":
        t = process_consequences(t)

    if vep_annotation not in t[vep_root].keys():
        raise ValueError(
            f"{vep_annotation} is not a row field of the VEP annotation in Table"
        )
    # Create top-level annotation for `vep_annotation` by pulling out from `vep_root`
    # annotation and explode the `vep_annotation`.
    if isinstance(t, hl.Table):
        t = t.transmute(**{vep_annotation: t[vep_root][vep_annotation]})
        t = t.explode(t[vep_annotation])
    else:
        t = t.transmute_rows(**{vep_annotation: t[vep_root][vep_annotation]})
        t = t.explode_rows(t[vep_annotation])

    return t
```

And one that is basically just to get the groupings so something like this:
```
def get_constraint_grouping_expr(
    vep_annotation_expr,
    coverage_expr = None,
    include_transcript_group: bool = True,
    include_canonical_group: bool = True,
):
    lof_expr = vep_annotation_expr.lof
    polyphen_prediction_expr = vep_annotation_expr.polyphen_prediction

    # Create constraint annotations to be used for groupings.
    groupings = {
        "annotation": vep_annotation_expr.most_severe_consequence,
        "modifier": hl.coalesce(lof_expr, polyphen_prediction_expr, "None"),
        "gene": vep_annotation_expr.gene_symbol,
    }
    if coverage_expr is not None:
        groupings["coverage"] = coverage_expr

    # Add 'transcript' and 'canonical' annotation if requested
    if include_transcript_group:
        groupings["transcript"] = vep_annotation_expr.transcript_id
    if include_canonical_group:
        groupings["canonical"] = hl.or_else(vep_annotation_expr.canonical == 1, False)

    return groupings
```

Then it adds some lines for their use, but I think it's better to split it like this to make it more generalizable.

Then add a function like this to the gnomad-constraint repo:
```
def annotate_constraint_groupings(ht, vep_annotation="transcript_consequences"):
    ht = explode_by_vep_annotation(ht, vep_annotation)
    if vep_annotation != "transcript_consequences":
        include_transcript_group = False
        include_canonical_group = False
    groupings = get_constraint_grouping_expr(
        ht[vep_annotation],
        coverage_expr=ht.exome_coverage,
        include_transcript_group=include_transcript_group,
        include_canonical_group=include_canonical_group,
    )

    return ht.annotate(**groupings), tuple(groupings.keys())
```